flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -862,11 +862,19 @@ private void onTriggerSuccess() {
      */
     private void onTriggerFailure(
             CheckpointTriggerRequest onCompletionPromise, Throwable throwable) {
-        final CheckpointException checkpointException =
-                getCheckpointException(
-                        CheckpointFailureReason.TRIGGER_CHECKPOINT_FAILURE, throwable);
-        onCompletionPromise.completeExceptionally(checkpointException);
-        onTriggerFailure((PendingCheckpoint) null, checkpointException);
+        Throwable rootThrowable = ExceptionUtils.stripCompletionException(throwable);
+        if (rootThrowable instanceof IOException) {
+            final CheckpointException checkpointException =
+                    getCheckpointException(CheckpointFailureReason.IO_EXCEPTION, rootThrowable);
+            onCompletionPromise.completeExceptionally(checkpointException);
+            onTriggerFailure((PendingCheckpoint) null, throwable);
+        } else {
+            final CheckpointException checkpointException =
+                    getCheckpointException(
+                            CheckpointFailureReason.TRIGGER_CHECKPOINT_FAILURE, throwable);
+            onCompletionPromise.completeExceptionally(checkpointException);
+            onTriggerFailure((PendingCheckpoint) null, checkpointException);
```

Comment:
I didn't get why `if` and `else` blocks are different(one passes throwable to onTriggerFailure while another passes checkpointException). As I understand you should only choose CheckpointFailureReason but another code should be the same. Or Did I understand something wrong?

Reply:
u r right and I have also made a change here.